internal/api/auth/ldap.go

```diff
@@ -1,41 +1,212 @@
+	ldapURL               string
+	bindDN                string
+	password              string
+	baseDN                string
+	maxConnectionLifetime time.Duration
 }
 
 func NewLDAPAuth(ldapURL string, bindDN string, password string, baseDN string, tlsConfig *tls.Config) *LDAPAuth {
-	conn, err := ldap.DialURL(ldapURL, ldap.DialWithTLSConfig(tlsConfig))
-	if err != nil {
-		log.Error().Err(err).Str("ldapURL", ldapURL).Msg("failed to connect to the LDAP server")
-		return nil
-	}
 	return &LDAPAuth{
-		ldapClient: conn,
-		bindDN:     bindDN,
-		password:   password,
-		baseDN:     baseDN,
+		tlsConfig: tlsConfig,
+		ldapURL:   ldapURL,
+		bindDN:    bindDN,
+		password:  password,
+		baseDN:    baseDN,
+		reqCh:     make(chan authRequest),
 	}
 }
 
+func SetLDAPDefaultTimeout(timeout time.Duration) {
+	ldap.DefaultTimeout = timeout //nolint:reassign  // we want to customize the default timeout
+}
+
+// SetMaxConnectionLifetime sets the maximum lifetime of a connection.
+//
+// The maximum lifetime is the maximum amount of time a connection may be reused for.
+// This is not a guarantee, as the connection may have been closed by the server before reaching that timer.
+func (l *LDAPAuth) SetMaxConnectionLifetime(maxConnectionLifetime time.Duration) {
+	l.maxConnectionLifetime = maxConnectionLifetime
+}
+
+// StartAuthenticationWorkers starts a pool of workers that will handle the authentication requests.
+func (l *LDAPAuth) StartAuthenticationWorkers(ctx context.Context, workersCount int) error {
+	if workersCount <= 0 {
+		return fmt.Errorf("'WorkersCount' must be greater than 0: %d", workersCount)
+	}
+	for i := 0; i < workersCount; i++ {
+		go l.spawnConnectionWorker(ctx)
+	}
+	return nil
+}
 func (l *LDAPAuth) AuthenticateUser(username string, password string) bool {
-	if err := l.ldapClient.Bind(l.bindDN, l.password); err != nil {
+	req := authRequest{
+		username: username,
+		password: password,
+		authResp: make(chan bool),
+	}
+	l.reqCh <- req
+	return <-req.authResp
+}
+
+func (l *LDAPAuth) spawnConnectionWorker(ctx context.Context) {
+	const maxRetry = 1
+	var conn *ldap.Conn
+	var err error
+	tick := time.NewTicker(l.maxConnectionLifetime)
+
+	for {
+		select {
+		case req := <-l.reqCh:
+			auth := false
+			retry := 0
+			for retry <= maxRetry+1 {
+				retry++
+				log.Debug().Msgf("worker LDAP authentication attempt number %d", retry)
+				// (re)connect if needed
+				if conn == nil || conn.IsClosing() {
+					log.Debug().Msg("LDAP connection is closed, reconnecting")
+					conn, err = l.connect()
+					if err != nil {
+						log.Error().Err(err).Msg("worker LDAP reconnection failed")
+						req.authResp <- false
+						break
+					}
+				}
+
+				// bind with the user credentials
+				var connState connectionStatus
+				ctxTimeout, cancel := context.WithTimeoutCause(ctx, ldap.DefaultTimeout, ErrLDAPTimeout)
+				auth, connState = l.authenticateWithTimeout(ctxTimeout, conn, req.username, req.password)
+				cancel()
```

Comment:
why not using `defer cancel()` between L115 and L116 ?

Reply:
as discussed internally, I wrapped it in an anonymous function and added the defer